metrics_handler/metrics.py

```diff
@@ -0,0 +1,246 @@
+        'Possible tags were: {}'.format(tag, raw_metrics.keys()))
+
+  # MetricPoints should be sorted by timestamp with earlier events first.
+  start_wall_time = values[0].wall_time
+  try:
+    end_wall_time = next(
+        v.wall_time for v in values
+        if v.metric_value >= threshold)
+    return MetricPoint(end_wall_time - start_wall_time, end_wall_time)
+  except StopIteration:
+    max_accuracy = max(v.metric_value for v in values)
+    raise ValueError(
+        'Accuracy metric `{}` was never high enough to satisfy the '
+        '`time_to_accuracy` settings from the config. Max accuracy: {}. '
+        'Target accuracy: {}. Config for `time_to_accuracy`: {}'.format(
+            tag, max_accuracy, threshold))
+
+
+def metric_bounds(value_history, threshold, comparison):
+  """Compute upper/lower bounds and whether metric is within those bounds.
+
+  Args:
+    value_history (list of floats): History of values for this metric. These
+      should be ordered so the most recent value is the last in the list.
+    threshold: Threshold, desired metric threshold.
+    comparison: string, comparison to given threshold.
+
+  Returns:
+    tuple(is_within_bounds (bool), lower_bound (float), upper_bound (float)).
+      lower_bound and/or upper_bound can be None.
+
+  Raises:
+    ValueError if the regression test config is invalid.
+  """
+  if not comparison or comparison not in ALLOWED_COMPARISONS:
+    raise ValueError(
+        'A metric success condition must set the `comparison` field to '
+        'one of {}. comparison was: {}'.format(
+            ALLOWED_COMPARISONS, comparison))
+
+  if threshold.type == 'fixed_value':
+    if 'greater' in comparison:
+      lower_bound = threshold.value
+      upper_bound = math.inf
+    elif 'less' in comparison:
+      lower_bound = -math.inf
+      upper_bound = threshold.value
+    elif comparison == 'equal':
+      lower_bound = threshold.value
+      upper_bound = threshold.value
+    else:
+      raise ValueError(
+          'A metric success condition using a `fixed_value`-type threshold '
+          'must use `greater`, `greater_or_equal`, `less`, `less_or_equal`, '
+          'or `equal` for the comparison. The comparison was: {}'.format(
+              comparison))
+  elif threshold.type == 'stddevs_from_mean':
+    if comparison not in ('greater', 'less', 'less_or_equal'):
+      raise ValueError(
+          'A metric success condition using a `stddevs_from_mean`-type '
+          'threshold must use `greater`, `less`, or `less_or_equal` for the '
+          'comparison. The comparison was: {}'.format(comparison))
+    values = [v.metric_value for v in value_history]
+    mean = np.mean(values)
+    stddev = np.std(values)
+
+    if 'less' in comparison:
+      lower_bound = -math.inf
+      upper_bound = mean + (stddev * threshold.value)
+    elif 'greater' in comparison:
+      lower_bound = max(mean - (stddev * threshold.value), 0)
```

Comment:
I copied this over from main.py. Is there a reason we can't allow values less than 0?

Reply:
It just affects how the bounds are rendered in the dashboard and I couldn't think of any metrics where it would be relevant to have negative numbers as bounds. That said, now I think it's better to remove this `max` and push any necessary changes for rendering onto the dashboard, e.g. the dashboard could replace negative values with 0 if that's what we want

Reply:
I agree that we can push that logic to the dashboard. Done.